### PR TITLE
Internal auth process in MS API Module

### PR DESCRIPTION
**Pull request by guyfreund**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Description
The auth process will be now determined by the args/params the user is providing to the ms graph integration.
The only thing the user has to do is to config the integration and MS API module will configure itself.
If building a new integration the developer should declare the following lines of code and that's it:
```
    args: dict = demisto.args()
    params: dict = demisto.params()
    self_deployed: bool = params.get('self_deployed', False)
    tenant_id: str = params.get('tenant_id', '')
    auth_and_token_url: str = params.get('auth_id', '')
    enc_key: str = params.get('enc_key', '')
    base_url: str = urljoin(params.get('url', ''), '/v1.0')
    app_name: str = 'ms-graph-mail'
    ok_codes: tuple = (200, 201, 202)
    use_ssl: bool = not params.get('insecure', False)
    proxy: bool = params.get('proxy', False)

    client: MsGraphClient = MsGraphClient(self_deployed, tenant_id, auth_and_token_url, enc_key, app_name, base_url,
                                          use_ssl, proxy, ok_codes)
```

## Minimum version of Demisto
- [ ] 4.5.0
- [ ] 5.0.0
- [ ] 5.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [ ] Documentation

### CODE REVIEW

**Comment by guyfreund** [comment]
For example, this was the configuration done in MS Graph Mail:
```
    """ COMMANDS MANAGER / SWITCH PANEL """
    command = demisto.command()
    args = demisto.args()
    LOG(f'Command being called is {command}')

    params = demisto.params()
    self_deployed = params.get('self_deployed', False)
    tenant_id = params.get('tenant_id')
    auth_and_token_url = params.get('auth_id', '')
    enc_key = params.get('enc_key')
    # Remove trailing slash to prevent wrong URL path to service
    url = params.get('url', '')
    server = url[:-1] if (url and url.endswith('/')) else url
    # Service base URL
    base_url = server + '/v1.0'
    app_name = 'ms-graph-mail'
    ok_codes = (200, 201, 202)
    use_ssl = not params.get('insecure', False)
    proxy = handle_proxy()

    if self_deployed:
        app_url = f'https://login.microsoftonline.com/{tenant_id}/oauth2/v2.0/token'
        ms_client = MicrosoftClient.from_self_deployed(tenant_id, auth_and_token_url,
                                                       enc_key, app_url=app_url,
                                                       scope='https://graph.microsoft.com/.default',
                                                       base_url=base_url, verify=use_ssl,
                                                       proxy=proxy, ok_codes=ok_codes)
    else:
        # params related to oproxy
        ms_client = MicrosoftClient.from_oproxy(auth_and_token_url, enc_key, app_name,
                                                tenant_id=tenant_id, base_url=base_url, verify=use_ssl,
                                                proxy=proxy, ok_codes=ok_codes)

    client = MsGraphClient(ms_client)
```

File: Packs/ApiModules/Scripts/MicrosoftApiModule/CHANGELOG.md [hunk 1]
```diff
@@ -1 +1,2 @@
-## [Unreleased]
\ No newline at end of file
+## [Unreleased]
+Implemented internal automated authentication.
```
Comment (Itay4) [inline Packs/ApiModules/Scripts/MicrosoftApiModule/CHANGELOG.md]: empty changelog

File: Integrations/MicrosoftGraphMail/MicrosoftGraphMail.py [hunk 2]
```diff
@@ -627,41 +627,31 @@ def get_email_as_eml_command(client: MsGraphClient, args):
...
+    params: dict = demisto.params()
+    self_deployed: bool = params.get('self_deployed', False)
+    tenant_id: str = params.get('tenant_id', '')
+    auth_and_token_url: str = params.get('auth_id', '')
+    enc_key: str = params.get('enc_key', '')
+    base_url: str = urljoin(params.get('url', ''), '/v1.0')
+    app_name: str = 'ms-graph-mail'
+    ok_codes: tuple = (200, 201, 202)
+    use_ssl: bool = not params.get('insecure', False)
+    app_url: str = f'https://login.microsoftonline.com/{tenant_id}/oauth2/v2.0/token' if self_deployed else ''
```
Comment (Itay4) [inline Integrations/MicrosoftGraphMail/MicrosoftGraphMail.py]: move to the module

File: Packs/ApiModules/Scripts/MicrosoftApiModule/MicrosoftApiModule.py [hunk 3]
```diff
@@ -15,13 +15,21 @@ class MicrosoftClient(BaseClient):
     def __init__(self, tenant_id: str = '', auth_id: str = '', enc_key: str = '',
                  token_retrieval_url: str = '', app_name: str = '', refresh_token: str = '',
                  client_id: str = '', client_secret: str = '', scope: str = '', resource: str = '', app_url: str = '',
-                 verify: bool = True, auth_type: str = OPROXY_AUTH_TYPE, *args, **kwargs):
+                 verify: bool = True, self_deployed: bool = False, *args, **kwargs):
         """
         Microsoft Client class that implements logic to authenticate with oproxy or self deployed applications.
         It also provides common logic to handle responses from Microsoft.
```
Comment (Itay4) [inline Packs/ApiModules/Scripts/MicrosoftApiModule/MicrosoftApiModule.py]: add all args to docstring and elaborate

File: Integrations/MicrosoftGraphMail/MicrosoftGraphMail.py [hunk 4]
```diff
@@ -627,41 +627,31 @@ def get_email_as_eml_command(client: MsGraphClient, args):
...
+    scope: str = 'https://graph.microsoft.com/.default' if self_deployed else ''
+
+    proxy: dict = handle_proxy()
```
Comment (Itay4) [inline Integrations/MicrosoftGraphMail/MicrosoftGraphMail.py]: change the logic to pass boolean

File: Integrations/MicrosoftGraphMail/MicrosoftGraphMail.py [hunk 5]
```diff
@@ -627,41 +627,31 @@ def get_email_as_eml_command(client: MsGraphClient, args):
...
+
+    ms_client = MicrosoftClient(self_deployed=self_deployed, tenant_id=tenant_id, auth_id=auth_and_token_url,
```
Comment (Itay4) [inline Integrations/MicrosoftGraphMail/MicrosoftGraphMail.py]: move this to the integration client

File: Integrations/MicrosoftGraphMail/CHANGELOG.md [hunk 6]
```diff
@@ -1,5 +1,7 @@
 ## [Unreleased]
  - Fixed an issue where the listing emails were not comparing the mail ID.
+ - Moved authentication to be handled in Microsoft API Module.
```
Comment (Itay4) [inline Integrations/MicrosoftGraphMail/CHANGELOG.md]: remove

File: Integrations/MicrosoftGraphListener/CHANGELOG.md [hunk 7]
```diff
@@ -1,4 +1,5 @@
 ## [Unreleased]
+Moved authentication to be handled in Microsoft API Module.
```
Comment (Itay4) [inline Integrations/MicrosoftGraphListener/CHANGELOG.md]: remove

File: Packs/ApiModules/Scripts/MicrosoftApiModule/MicrosoftApiModule.py [hunk 8]
```diff
@@ -14,70 +14,37 @@ class MicrosoftClient(BaseClient):
...
+        if not self_deployed:
+            auth_id_and_token_retrieval_url = auth_id.split('@')
+            auth_id = auth_id_and_token_retrieval_url[0]
+            if len(auth_id_and_token_retrieval_url) != 2:
+                token_retrieval_url = 'https://oproxy.demisto.ninja/obtain-token'  # guardrails-disable-line
+            else:
+                token_retrieval_url = auth_id_and_token_retrieval_url[1]
+        else:
+            self.app_url = f'https://login.microsoftonline.com/{tenant_id}/oauth2/v2.0/token'
+            self.scope = 'https://graph.microsoft.com/.default'
```
Comment (Itay4) [inline Packs/ApiModules/Scripts/MicrosoftApiModule/MicrosoftApiModule.py]: add this as constructor arg with that value as default

File: Packs/ApiModules/Scripts/MicrosoftApiModule/MicrosoftApiModule.py [hunk 9]
```diff
@@ -14,70 +14,37 @@ class MicrosoftClient(BaseClient):
...
-            app_url: The self deployed application request URL
-        Returns:
-            An instance of Microsoft Client with self deployed application authentication.
-        """
-        return cls(tenant_id=tenant_id, client_id=client_id,  # type: ignore[misc]
-                   client_secret=client_secret, auth_type=SELF_DEPLOYED_AUTH_TYPE,
-                   scope=scope, resource=resource, app_url=app_url, *args, **kwargs)
-
-    def http_request(self, *args, **kwargs) -> requests.Response:
+    def http_request(self, *args, **kwargs):
```
Comment (Itay4) [inline Packs/ApiModules/Scripts/MicrosoftApiModule/MicrosoftApiModule.py]: figure the return type hint

File: Packs/ApiModules/Scripts/MicrosoftApiModule/MicrosoftApiModule.py [hunk 10]
```diff
@@ -14,69 +14,45 @@ class MicrosoftClient(BaseClient):
...
                  token_retrieval_url: str = '', app_name: str = '', refresh_token: str = '',
-                 client_id: str = '', client_secret: str = '', scope: str = '', resource: str = '', app_url: str = '',
-                 verify: bool = True, auth_type: str = OPROXY_AUTH_TYPE, *args, **kwargs):
+                 scope: str = 'https://graph.microsoft.com/.default',
+                 resource: str = '', verify: bool = True, self_deployed: bool = False, *args, **kwargs):
         """
         Microsoft Client class that implements logic to authenticate with oproxy or self deployed applications.
         It also provides common logic to handle responses from Microsoft.
+        Args:
+            tenant_id: if self deployed then it's the tenant for the app url, or if oproxy then it's the token
```
Comment (Itay4) [inline Packs/ApiModules/Scripts/MicrosoftApiModule/MicrosoftApiModule.py]: IMPROVE ENGLISH

File: Packs/ApiModules/Scripts/MicrosoftApiModule/MicrosoftApiModule.py [hunk 11]
```diff
@@ -14,69 +14,45 @@ class MicrosoftClient(BaseClient):
...
+            auth_id: if self deployed then it's the client id, or if oproxy then it's the auth id and token url
```
Comment (Itay4) [inline Packs/ApiModules/Scripts/MicrosoftApiModule/MicrosoftApiModule.py]: optionally.. write about dev..

File: Integrations/MicrosoftGraphListener/MicrosoftGraphListener.py [hunk 12]
```diff
@@ -793,35 +811,31 @@ def main():
 
     # params related to common instance configuration
     base_url = 'https://graph.microsoft.com/v1.0/'
-    verify = not params.get('insecure', False)
+    use_ssl = not params.get('insecure', False)
     proxy = params.get('proxy', False)
     ok_codes = (200, 201, 202)
+    refresh_token = params.get('refresh_token', '')
+    auth_and_token_url: str = params.get('auth_id', '')
```
Comment (Itay4) [inline Integrations/MicrosoftGraphListener/MicrosoftGraphListener.py]:
```suggestion
    auth_and_token_url = params.get('auth_id', '')
```

File: Integrations/MicrosoftGraphListener/MicrosoftGraphListener.py [hunk 13]
```diff
@@ -793,35 +796,31 @@ def main():
...
-        refresh_token = (demisto.getIntegrationContext().get('current_refresh_token')
-                         or params.get('refresh_token', ''))
-        enc_key = params.get('enc_key')
-        app_name = 'ms-graph-mail-listener'
-        ms_client = MicrosoftClient.from_oproxy(params.get('auth_id', ''), enc_key, app_name,
-                                                refresh_token=refresh_token,
-                                                base_url=base_url, verify=verify, proxy=proxy, ok_codes=ok_codes)
-    client = MsGraphClient(ms_client, mailbox_to_fetch, folder_to_fetch, first_fetch_interval, emails_fetch_limit)
+    # params related to self deployed
+    tenant_id = refresh_token if self_deployed else ''
```
Comment (Itay4) [inline Integrations/MicrosoftGraphListener/MicrosoftGraphListener.py]: did we test if we can remove this ?

Comment (guyfreund) [reply Integrations/MicrosoftGraphListener/MicrosoftGraphListener.py]: We can't test it since self deployed isn't working.

**Pull request merged**